.teamcity/Gradle_Check/model/CIBuildModel.kt

```diff
@@ -192,20 +195,118 @@ data class CIBuildModel(
         GradleSubproject("performance", unitTests = false, functionalTests = false),
         GradleSubproject("runtimeApiInfo", unitTests = false, functionalTests = false),
         GradleSubproject("smokeTest", unitTests = false, functionalTests = false))
-)
+) {
+    val buildTypeBuckets: List<BuildTypeBucket>
 
-data class GradleSubproject(val name: String, val unitTests: Boolean = true, val functionalTests: Boolean = true, val crossVersionTests: Boolean = false, val containsSlowTests: Boolean = false) {
-    fun asDirectoryName(): String {
-        return name.replace(Regex("([A-Z])"), { "-" + it.groups[1]!!.value.toLowerCase() })
+    init {
+        val subprojectMap = subProjects.map { it.name to it }.toMap()
+        val buckets = listOf(
+            listOf("resources", "resourcesGcs", "resourcesHttp", "resourcesS3", "resourcesSftp"),
+            listOf("platformBase", "platformJvm", "platformNative")
+        )
+        val largeSubprojects = mapOf("integTest" to 3, "core" to 3, "dependencyManagement" to 3)
+
+        buildTypeBuckets =
+            listOf<BuildTypeBucket>(
+                SubprojectBucket(name = "AllUnitTest", subprojects = subProjects.filter { it.hasOnlyUnitTests() })
+            ) + buckets.map { projectsInBucket ->
+                SubprojectBucket(name = projectsInBucket[0], subprojects = projectsInBucket.map { subprojectMap.getValue(it) })
+            } + largeSubprojects.entries.map { entry ->
+                (1..entry.value).map { SubprojectSplit(subproject = subprojectMap.getValue(entry.key), number = it, total = entry.value) }
+            }.flatten() + subProjects.filter {
+                !buckets.flatten().contains(it.name) && !largeSubprojects.containsKey(it.name) && !it.hasOnlyUnitTests()
```

Comment:
I still think it would be better to first accumulate all the non-trivial buckets, and then add all the projects which are not already present in that buckets. Then there wouldn't be a need to to this if here and it would be more future proof.